### ISSUE TITLE
Nuget packages not properly restored when ReadyToRun publishing

### ISSUE THREAD

**Issue by alex-jitbit** [dotnet/runtime#64260]
### Description

After upgrading nuget System.Data.SqlClient from 4.8.2 to 4.8.3 I get the above error on production: `Could not load file or assembly 'System.Data.SqlClient, Version=4.6.1.3'`

Works fine on local machine running under Visual Studio IIS Express but after PUBLISHING a "ReadyToRun" win-x64 app to a folder and then copying to the server - it stops working. 4.8.2 worked fine, it's a regression.

The weird thing is that both 4.8.3 and 4.8.2 publish the same System.Data.SqlClient.dll identical file that is 2-something megabytes large. I guess that is the problem - publishing the wrong files. Must be the "ReadyToRun" publishing messes up versioning somehow?

My guess is "ReadyToRun" publishing has some internal caching logic and does not "recompile" the file because it's not detecting any changes after moving from 4.8.2 to 4.8.3

Clearing the solution, deleting the "obj" folder before publishing resolves the problem

### Reproduction Steps

Publish a readyToRun ASP.NET 5 Core app that references SqlClient 4.8.2 in VS 2019 to a folder and try to run it. Works fine with 4.8.2. Then upgrade Nuget SqlClient to 4.8.3 and republish with readytorun - it stops working.

### Expected behavior

Should be working

### Actual behavior

Could not load file or assembly 'System.Data.SqlClient, Version=4.6.1.3'

### Known Workarounds

Clearing the solution and deleting obj folders prior to publishing helps.

**Comment by alex-jitbit** [comment]
Just to clarify: SqlClient is just an example here. It's not the only package that causes this. For instance BouncyCastle.Crypto surfaces the same issue, as long as some other packages.

**Comment by ltrzesniewski** [comment]
I think you're experiencing the same issue as one I reported a while back: https://github.com/dotnet/sdk/issues/12379

**Comment by alex-jitbit** [comment]
@ltrzesniewski yeah, looks like so. And the workaround looks similar too (delete all the cache in `obj` folder and republish)

**Comment by alex-jitbit** [comment]
Here's the fix we use, very similar to @ltrzesniewski 's

```xml
  <Target Name="DeleteR2RCache" BeforeTargets="BeforePublish">
    <Message Text="Clearing ReadyTpRun cache" Importance="High" />
    <RemoveDir Directories="$(IntermediateOutputPath)R2R" />
  </Target>
 ```
 
Put it in your Publish Profile `.pubxml` file

**Comment by alex-jitbit** [comment]
@mangod9 one of the two can be closed for sure. Either this or https://github.com/dotnet/sdk/issues/12379
